Expand do loops with cycle and exit

Irvise opened the pull request:
This is a minor expansion of the tutorial page on control flow.

I am not entirely happy with the current result, but I cannot figure out why (probably because it is very short when compared to the rest of the file). So any comments on these changes are welcome.

I expect to continue expanding this with `do concurrent` and tags. However, some initial feedback will be beneficial.

This pull is related to the discussion in #82.

Regards,

Fernando

LKedward commented:
This looks great @Irvise, definitely an improvement!
Thanks 👍

certik commented:
Thanks @Irvise, looks great!

milancurcic commented:
I agree, great addition, thank you!

#build_preview

Irvise commented:
I suppose this marks the end of the changes.

Please, check `do concurrent` since I am not as well versed with it and I may have committed some mistakes.

More comments are welcome.

LKedward commented:
Apologies for the delay @Irvise, I will review this today with specific feedback.

LKedward commented on learn/quickstart/operators_control_flow.md:
```diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+
+### `do concurrent()`
+
+The `do concurrent` loop is used to specify that the _inside of the loop has no interdependencies_. By this, it is indicated
+that any loop does not depend on the loops before it. It is also necessary that any state changes that may occur must only
+happen within each `do concurrent` loop. Because of these requirements, it is important to be carful with what is written inside the loop.
```
```suggestion
The `do concurrent` loop is used to explicitly specify that the _inside of the loop has no interdependencies_; this informs the compiler that it may use parallelization/_SIMD_ to speed-up execution of the loop and conveys programmer intention more clearly. More specifically, this means
that any loop iteration does not depend on the prior execution of other loop iterations. It is also necessary that any state changes that may occur must only happen within each `do concurrent` loop. 
These requirements place restrictions on what can be placed within the loop body.
```

LKedward commented on learn/quickstart/operators_control_flow.md:
```diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+that any loop does not depend on the loops before it. It is also necessary that any state changes that may occur must only
+happen within each `do concurrent` loop. Because of these requirements, it is important to be carful with what is written inside the loop.
+
+However, the true nature of the `do concurrent` construct may not be fully understood by the paragraph above. `do concurrent`
+was specifically created to tell the compiler that it may parallelize/_SIMD_ what is inside the `do concurrent`,
+which could potentially lead to faster code. The intention of the programmer is also conveyed more clearly.
```
```suggestion
```

I've moved this information to the first paragraph in my previous suggestion - I think the motivation for `do concurrent` is better stated upfront.

LKedward commented on learn/quickstart/operators_control_flow.md:
````diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
   end do
 ```
 
+### `do while()`
````
```suggestion
### Conditional loop (`do while`)
```

LKedward commented on learn/quickstart/operators_control_flow.md:
````diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+  ! Here i = 11
+```
+
+### `do concurrent()`
````
```suggestion
### Parallelizable loop (`do concurrent`)
```

LKedward commented on learn/quickstart/operators_control_flow.md:
````diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+  print *, result_sin
+```
+
+### Finer loop control: (`exit`) and (`cycle`)
````
```suggestion
### Loop control statements (`exit` and `cycle`)
```

LKedward commented on learn/quickstart/operators_control_flow.md:
````diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+    print *, i
+  end do
+```
+### Even finer loop control: tags
````
```suggestion
{% include note.html content="When used within nested loops, the `cycle` and `exit` statements operate on the inner-most loop."}

### Nested loop control: tags
```

LKedward commented on learn/quickstart/operators_control_flow.md:
```diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+  outer_loop: do i=1,10
+    inner_loop: do j=1,10
+      if((j+i) .gt. 10) then ! Print only pairs of i and j that add up to 10
+	    cycle outer_loop     ! Go to the next iteration of the outer loop
+	  end if
+	  print *, 'I=', i, ' J=', j, ' Sum=', j+i
```
```suggestion
	  cycle outer_loop     ! Go to the next iteration of the outer loop
	end if
	print *, 'I=', i, ' J=', j, ' Sum=', j+i
```

LKedward commented on learn/quickstart/operators_control_flow.md:
```diff
@@ -119,19 +119,104 @@ __Example:__ `do` loop with skip
+
+### Finer loop control: (`exit`) and (`cycle`)
+
+Most often than not, loops need to be stopped if a condition is met. Fortran presents two reserved words to deal
```
```suggestion
Most often than not, loops need to be stopped if a condition is met. Fortran provides two executable statements to deal
```

LKedward commented:
#build_preview

LKedward commented:
#build_preview

LKedward commented:
#build_preview

Irvise commented:
I have moved  `do concurrent` to the bottom and I have updated some comparison symbols to a more "modern" standard.

I would say LGTM. Thanks @LKedward for the improvements!

LKedward commented:
Great stuff! 👍  (Sorry for the conflicts!)

LKedward commented:
Great stuff! 👍  (Sorry for the conflicts!)
#build_preview

LKedward commented on learn/quickstart/operators_control_flow.md:
```diff
@@ -119,19 +119,103 @@ __Example:__ `do` loop with skip
+
+{% include important.html content="`do concurrent` is not a basic feature of Fortran. The explanation given does not detail
+all the requirements that need to be met in order to write a correct `do concurrent` loop. Compilers are also free to do as they see fit,
+which means they may not optimize the loop."}
```
```suggestion
which means they may not optimize the loop." %}
```

This is my fault I think, I edited on your remote while you were still working @Irvise.
I will put as a suggestion for you to commit so not to cause any more conflicts.

Irvise commented:
#build_preview

LKedward commented:
One last thing @Irvise, some of the indentation is rendering incorrectly for the cycle example, and the tagged nested loop example; the `end if` is not aligned with the opening `if` and the loop body has extra indentation. Can you make sure you're not using any `tab` characters, only `spaces`?

Irvise commented:
#build_preview

LKedward commented:
I will merge as no other comments. Great addition @Irvise, thanks!

LKedward commented:
#delete_preview
